### PR TITLE
Add documentation about performance PRs, add (TBD) section on feature criteria

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/datafusion/issues/12357

## Rationale for this change

I think there is some "common knowledge" that is not explicitly written down anywhere

Specifically, that performance improvements should be accompanied by benchmark results (most recently this came up in  https://github.com/apache/datafusion/pull/12270)

## What changes are included in this PR?

1. Document need for benchmarks for performance PRs 
2. Add a section (without content) for the outcome of https://github.com/apache/datafusion/issues/12357
3. Some other "drive by" documentation improvements

Note I am not trying to change anything, only document the current reality.

## Are these changes tested?

by doc CI

## Are there any user-facing changes?
Docs

### CODE REVIEW

File: docs/source/contributor-guide/index.md [hunk 1]
```diff
@@ -88,35 +106,61 @@ committer who approved your PR to help remind them to merge it.
...
 
 Please ensure any issues you raise contains a rationale and suggested alternative -- it is frustrating to be told "don't do it this way" without any clear reason or alternate provided.
 
 Some things to specifically check:
 
-1. Is the feature or fix covered sufficiently with tests (see `Test Organization` below)?
+1. Is the feature or fix covered sufficiently with tests (see the [Testing](testing.md) section)?
 2. Is the code clear, and fits the style of the existing codebase?
 
+## Performance Improvements
```
Comment (alamb) [inline docs/source/contributor-guide/index.md:138]: This is the new content

File: docs/source/contributor-guide/index.md [hunk 2]
```diff
@@ -88,35 +106,61 @@ committer who approved your PR to help remind them to merge it.
...
 
-When reviewing PRs, please remember our primary goal is to improve DataFusion and its community together. PR feedback should be constructive with the aim to help improve the code as well as the understanding of the contributor.
+When reviewing PRs, our primary goal is to improve DataFusion and its community together. PR feedback should be constructive with the aim to help improve the code as well as the understanding of the contributor.
 
 Please ensure any issues you raise contains a rationale and suggested alternative -- it is frustrating to be told "don't do it this way" without any clear reason or alternate provided.
 
 Some things to specifically check:
 
-1. Is the feature or fix covered sufficiently with tests (see `Test Organization` below)?
+1. Is the feature or fix covered sufficiently with tests (see the [Testing](testing.md) section)?
```
Comment (alamb) [inline docs/source/contributor-guide/index.md:135]: added link (there is actually no `Testing Organization` section anymore below)

File: docs/source/contributor-guide/index.md [hunk 3]
```diff
@@ -88,35 +106,61 @@ committer who approved your PR to help remind them to merge it.
...
+To help committers evaluate the potential improvement, performance PRs should
+in general be accompanied by benchmark results that demonstrate the improvement.
+
+The best way to demonstrate a performance improvement is with the existing
+benchmarks:
+
+- [System level SQL Benchmarks](https://github.com/apache/datafusion/tree/main/benchmarks)
+- Microbenchmarks such as those in [functions/benches](https://github.com/apache/datafusion/tree/main/datafusion/functions/benches)
+
+If there is no suitable existing benchmark, you can create a new one. It helps
```
Comment (comphead) [inline docs/source/contributor-guide/index.md:157]: This is a good point, very often reviewing PRs we have to ask showing the actual performance benefit with existing benches or other performance tests

**Review by comphead — APPROVED** [review]
lgtm, thanks @alamb it looks more clear now

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @alamb This change makes the guide more friendly, especially for newcomers.

**Pull request merged**